File: app-tracker-automation/components/table.py
"""
Table component for handling data tables.
"""

from playwright.sync_api import Page
from components.base_component import BaseComponent


class Table(BaseComponent):
    """Table component class"""
    
    def __init__(self, page: Page, locator: str):
        super().__init__(page, locator)
    
    def get_row_count(self) -> int:
        """Get number of rows in table"""
        rows = self.element.locator("tbody tr")
        return len(rows.all())
    
    def get_column_count(self) -> int:
        """Get number of columns in table"""
        headers = self.element.locator("thead th")
        return len(headers.all())
    
    def get_cell_text(self, row_index: int, column_index: int) -> str:
        """Get text from specific cell"""
        cell = self.element.locator(f"tbody tr:nth-child({row_index + 1}) td:nth-child({column_index + 1})")
        return cell.text_content()
    
    def get_row_data(self, row_index: int) -> list:
        """Get all data from specific row"""
        row = self.element.locator(f"tbody tr:nth-child({row_index + 1})")
        cells = row.locator("td")
        return cells.all_inner_texts()
    
    def get_column_data(self, column_index: int) -> list:
        """Get all data from specific column"""
        cells = self.element.locator(f"tbody tr td:nth-child({column_index + 1})")
        return cells.all_inner_texts()
    
    def get_all_data(self) -> list:
        """Get all table data as 2D list"""
        rows = self.element.locator("tbody tr")
        all_data = []
        for i in range(len(rows.all())):
            all_data.append(self.get_row_data(i))
        return all_data
    
    def find_row_by_text(self, search_text: str) -> int:
        """Find row index containing specific text"""
        rows = self.element.locator("tbody tr")
        for i, row in enumerate(rows.all()):
            if search_text in row.text_content():
                return i
        return -1
    
    def click_row(self, row_index: int):
        """Click on specific row"""
        row = self.element.locator(f"tbody tr:nth-child({row_index + 1})")
        row.click()
    
    def sort_by_column(self, column_index: int):
        """Sort table by specific column"""
        header = self.element.locator(f"thead th:nth-child({column_index + 1})")
        header.click()
    
    def is_row_selected(self, row_index: int) -> bool:
        """Check if specific row is selected"""
        row = self.element.locator(f"tbody tr:nth-child({row_index + 1})")
        return "selected" in row.get_attribute("class")

File: ts-framework/components/table.ts
import { Page, Locator, expect } from '@playwright/test';
import { BaseComponent } from './base_component';

/**
 * Table component for handling data tables.
 */
export class Table extends BaseComponent {
    constructor(page: Page, locator: string) {
        super(page, locator);
    }

    async getRowCount(): Promise<number> {
        const rows = await this.element.locator("tbody tr").all();
        return rows.length;
    }

    async getColumnCount(): Promise<number> {
        const headers = await this.element.locator("thead th").all();
        return headers.length;
    }

    async getCellText(rowIndex: number, columnIndex: number): Promise<string | null> {
        const cell = this.element.locator(`tbody tr:nth-child(${rowIndex + 1}) td:nth-child(${columnIndex + 1})`);
        return await cell.textContent();
    }

    async getRowData(rowIndex: number): Promise<string[]> {
        const row = this.element.locator(`tbody tr:nth-child(${rowIndex + 1})`);
        const cells = row.locator("td");
        return await cells.allInnerTexts();
    }

    async getColumnData(columnIndex: number): Promise<string[]> {
        const cells = this.element.locator(`tbody tr td:nth-child(${columnIndex + 1})`);
        return await cells.allInnerTexts();
    }

    async getAllData(): Promise<string[][]> {
        const rowCount = await this.getRowCount();
        const allData: string[][] = [];
        for (let i = 0; i < rowCount; i++) {
            allData.push(await this.getRowData(i));
        }
        return allData;
    }

    async findRowByText(searchText: string): Promise<number> {
        const rows = await this.element.locator("tbody tr").all();
        for (let i = 0; i < rows.length; i++) {
            const text = await rows[i].textContent();
            if (text && text.includes(searchText)) {
                return i;
            }
        }
        return -1;
    }

    async clickRow(rowIndex: number): Promise<void> {
        const row = this.element.locator(`tbody tr:nth-child(${rowIndex + 1})`);
        await row.click();
    }

    async sortByColumn(columnIndex: number): Promise<void> {
        const header = this.element.locator(`thead th:nth-child(${columnIndex + 1})`);
        await header.click();
    }

    async isRowSelected(rowIndex: number): Promise<boolean> {
        const row = this.element.locator(`tbody tr:nth-child(${rowIndex + 1})`);
        const className = await row.getAttribute("class") || "";
        return className.includes("selected");
    }
}

/**
 * App Tracker Specific Table Selectors
 */
export class AppTrackerTable extends Table {
    static readonly APPLICATION_LIST_CONTAINER = ".MuiBox-root.jss137.application-listing-container";
    static readonly APPLICATION_ROW = ".MuiBox-root.jss138";
    static readonly APPLICATION_NO = ".application-no-val";
    static readonly USER_NAME = ".uname-val";
    static readonly PLAN_NAME = ".plan-name";

    constructor(page: Page) {
        super(page, AppTrackerTable.APPLICATION_LIST_CONTAINER);
    }

    async getApplicationNumber(rowIndex: number = 0): Promise<string | null> {
        const row = this.page.locator(AppTrackerTable.APPLICATION_ROW).nth(rowIndex);
        return await row.locator(AppTrackerTable.APPLICATION_NO).textContent();
    }

    async getUserName(rowIndex: number = 0): Promise<string | null> {
        const row = this.page.locator(AppTrackerTable.APPLICATION_ROW).nth(rowIndex);
        return await row.locator(AppTrackerTable.USER_NAME).textContent();
    }

    async getPlanName(rowIndex: number = 0): Promise<string | null> {
        const row = this.page.locator(AppTrackerTable.APPLICATION_ROW).nth(rowIndex);
        return await row.locator(AppTrackerTable.PLAN_NAME).textContent();
    }

    async clickApplicationRow(rowIndex: number = 0): Promise<void> {
        const row = this.page.locator(AppTrackerTable.APPLICATION_ROW).nth(rowIndex);
        await row.click();
    }
}
